File: src/main/scala/scaled/java/PropertiesMode.scala
//
// Scaled Java Mode - a Scaled major mode for editing Java code
// http://github.com/scaled/java-mode/blob/master/LICENSE

package scaled.java

import scaled._
import scaled.grammar._
import scaled.code.{CodeConfig, Commenter}

object PropertiesConfig extends Config.Defs {
  import CodeConfig._
  import GrammarConfig._

  // map TextMate grammar scopes to Scaled style definitions
  val effacers = List(
    effacer("comment.line", commentStyle),
    effacer("keyword", keywordStyle)
  )

  // map TextMate grammar scopes to Scaled syntax definitions
  val syntaxers = List(
    syntaxer("comment.line", Syntax.LineComment)
  )

  def propsGrammar = Grammar.parseNDF(stream("JavaProperties.ndf"))
  lazy val grammars = Seq(propsGrammar)
}

@Major(name="properties",
       tags=Array("code", "project", "properties"),
       pats=Array(".*\\.properties", "package.scaled", "module.scaled"),
       desc="A major mode for editing Java properties files.")
class PropertiesMode (env :Env) extends GrammarCodeMode(env) {

  override def configDefs = PropertiesConfig :: super.configDefs
  override def grammars = PropertiesConfig.grammars
  override def effacers = PropertiesConfig.effacers
  override def syntaxers = PropertiesConfig.syntaxers

  override val indenters = Nil
  override val commenter = new Commenter() {
    override def linePrefix = "#"
    // TODO: ! is also a comment start character, sigh...
  }
}

File: src/main/scala/scaled/java/PropertiesMode.java
//
// Scaled Java Mode - a Scaled major mode for editing Java code
// http://github.com/scaled/java-mode/blob/master/LICENSE

package scaled.java;

import scala.collection.immutable.List;
import scala.collection.immutable.Nil$;
import scala.collection.immutable.Seq;

import scaled.*;
import scaled.code.CodeConfig;
import scaled.code.Commenter;
import scaled.code.Indenter;
import scaled.grammar.*;
import static scaled.code.CodeConfig.*;

@Major(name="properties",
       tags={ "code", "project", "properties" },
       pats={ ".*\\.properties", "package.scaled", "module.scaled" },
       desc="A major mode for editing Java properties files.")
public class PropertiesMode extends GrammarCodeMode {

  public PropertiesMode (Env env) { super(env); }

  @Override public List<Config.Defs> configDefs () {
    return SC.cons(PropertiesConfig.INSTANCE, super.configDefs());
  }

  @Override public Seq<Grammar> grammars () {
    return PropertiesConfig.INSTANCE.grammars;
  }
  @Override public List<Selector.Fn> effacers () {
    return PropertiesConfig.INSTANCE.effacers;
  }
  @Override public List<Selector.Fn> syntaxers () {
    return PropertiesConfig.INSTANCE.syntaxers;
  }

  @Override public List<Indenter> indenters () {
    return SC.nil();
  }
  @Override public Commenter commenter () {
    return new Commenter() {
      // TODO: ! is also a comment start character, sigh...
      @Override public String linePrefix () { return "#"; }
    };
  }
}
